torch/csrc/jit/codegen/cuda/scheduler/normalization.cpp

```diff
@@ -334,20 +338,25 @@ std::shared_ptr<ReductionParams> innerPersistentHeuristic(
 
   // Try moving persistent buffer factors into threads until we have too many
   // threads.
+  constexpr int batches_per_block_inner_reduction_max = 10;
   while (
-      // If using less than a quarter of available threads
-      bdimx * bdimy * bdimz * 2 <=
-          ceilDiv(device_max_threads_per_multiprocessor, (int64_t)4) &&
+      // If block size can be doubled
+      bdimx * bdimy * bdimz * 2 <= max_threads_in_block &&
       // And batches_per_block_inner_reduction can be divided by two
-      (batches_per_block_inner_reduction >= 2 ||
+      (batches_per_block_inner_reduction >=
+           batches_per_block_inner_reduction_max ||
        batches_per_block_outer_reduction >= 2)) {
     // Try to decrease per thread register allocation persistence size on inner
     // reduction
-    if (batches_per_block_inner_reduction >= 2 &&
-        batches_per_block_inner_reduction !=
-            roundUpPow2Or8(batches_per_block_inner_reduction / 2)) {
-      batches_per_block_inner_reduction =
-          roundUpPow2Or8(batches_per_block_inner_reduction / 2);
+    if (batches_per_block_inner_reduction >=
+        batches_per_block_inner_reduction_max) {
+      if (batches_per_block_inner_reduction * bdimx ==
+          threads_after_vectorize) {
```

Comment:
I don't understand what this condition means. What does it imply?

Reply:
It means remaining reductions are uniformly distributed between `bdimx` and `batches_per_block_inner_reduction`. e.g. case 40960, `threads_after_vectorize= 40960/8=5120`, `batches_per_block_inner_reduction=40`, and `bdimx=128`. After this piece of code, `batches_per_block_inner_reduction=40/2/2=10`, and `bdimx=128*2*2=512`

Reply:
OK, please make sure to add comments, or even better if code is self explanatory. The previous commit was a great example of self explanatory code as it had a function like `isEvenlyDistributed`.

Reply:
Hmm, this condition does not mean the new `batched_per_block_inner_reduction` and `bdimx` as updated below would have uniformly distributed work, right? Shouldn't the new values be considered instead?

Reply:
It seems this check is completely gone in the latest commit. Why is it removed? Don't you want to make sure the uniformness?

Reply:
reason is `batched_per_block_inner_reduction` will be reduced by half whether distribution is uniform or not. In most cases, the distribution is uniform as we are staring from 128 threads. If not uniform, the difference between reduce by half and reduce by one is <5%.